### ISSUE TITLE
bastion outbound security group

### ISSUE THREAD

**Issue by davidaurban** [joshuamkite/terraform-aws-ssh-bastion-service#61]
The outbound security group is too permissive for us.  Is there a way to replace the outbound?  We can only have 53, 80 and 443.  Thanks.

**Comment by joshuamkite** [comment]
From the Readme - [Notes for deployment](https://github.com/joshuamkite/terraform-aws-ssh-bastion-service#notes-for-deployment)

> You can supply list of one or more security groups to attach to the host instance launch configuration within the module if you wish. This can be supplied together with or instead of a whitelisted range of CIDR blocks. It may be useful in an enterprise setting to have security groups with rules managed separately from the bastion plan but of course if you do not assign either a suitable security group or whitelist then you may not be able to reach the service!

**Comment by joshuamkite** [comment]
OK, apologies, I think we have both misunderstood things here. 

1. We don't need to change the launch configuration to apply additional security groups, e.g.  this works (based on `example/full_with_public_ip`):

```terraform
module "ssh-bastion-service" {
  source = "joshuamkite/ssh-bastion-service/aws"
  # source                        = "../../"
  aws_region       = var.aws_region
  environment_name = var.environment_name
  vpc              = aws_vpc.bastion.id
  subnets_asg      = flatten([aws_subnet.bastion.*.id])
  subnets_lb       = flatten([aws_subnet.bastion.*.id])
  # cidr_blocks_whitelist_service = [var.everyone_cidr]
  security_groups_additional = [aws_security_group.custom.id]
  public_ip                  = true
  depends_on = [
    aws_vpc.bastion,
    aws_subnet.bastion,
    aws_internet_gateway.bastion,
  ]
  bastion_instance_types = ["t3.micro"]
}

resource "aws_security_group" "custom" {
  name                   = "custom"
  description            = "custom security group"
  revoke_rules_on_delete = true
  vpc_id                 = aws_vpc.bastion.id
  tags                   = var.tags
}

resource "aws_security_group_rule" "custom" {
  type              = "ingress"
  from_port         = 22
  to_port           = 22
  protocol          = "tcp"
  cidr_blocks       = var.custom_cidr
  security_group_id = aws_security_group.custom.id
  description       = "custom security group rule"
}

variable "custom_cidr" {
  default= ["0.0.0.0/32"]
} 
```
2. You are however looking to implement a more restrictive _outgoing_ policy. At the moment this is hard-coded with (in `security_group.tf`):

```terraform
# Permissive egress policy because we want users to be able to install their own packages 

resource "aws_security_group_rule" "bastion_host_out" {
  type              = "egress"
  from_port         = 0
  to_port           = 65535
  protocol          = -1
  security_group_id = aws_security_group.bastion_service.id
  cidr_blocks       = ["0.0.0.0/0"]
  description       = "bastion service and host egress"
}
```

There are a couple of things to consider here:

1. The image is built at run time and installs packages via `apt`; `docker` and `go`. Ports 53; 80; 443 _probably_ would cover these but it would be a new feature and require testing.
2. If you're serious about limiting access on your VPC then you should be using NACLs. Whilst these are more tricky to use they're a lot more assured than relying on security groups attached to individual resources alone. Would this be an option for you?

**Comment by joshuamkite** [comment]
I have had a brief play looking into this. It's pretty straightforward to put a ternary evaluation on creating the outgoing security group rule and have egress managed via the additional security group. The issues then are:

- Does the container build? (untested)
- We also need to ensure sshd in the container is reachable, so:
   - substitute our own `docker_setup.tftpl` which sets ssh in the container to run on a reachable port, e.g. 443 in your case
   - substitute our own `systemd.tftpl` which sets the systemd templated unit on the host to run on a reachable port, e.g. 443 in your case

This is a fair amount of work and whilst interesting may take some time to implement alone...

**Comment by joshuamkite** [comment]
Please review [ feature/custom_outbound_security_group](https://github.com/joshuamkite/terraform-aws-ssh-bastion-service/tree/feature/custom_outbound_security_group)

**Comment by joshuamkite** [comment]
Fixed with [release 8.1.0](https://github.com/joshuamkite/terraform-aws-ssh-bastion-service/tree/8.1.0)
